llvm-external-projects/iree-dialects/lib/Dialect/LinalgExt/Passes/ConvertConv2DToWinograd.cpp

```diff
@@ -0,0 +1,388 @@
+/// function computes matmul(G, matmul(F, transpose(G))) where
+/// F is a tile of the convolution filter of size m x m
+/// (single input channel, single output channel) and G has
+/// shape m x (m + r - 1) where r is the output tile size and
+/// (m + r - 1) is the input tile size.
+/// The time complexity of this function is O(ic * oc)
+/// where ic is the number of input channels and oc is the
+/// number of output channels since input tile size and kernel size
+/// are constants. So for large ic and oc, this function is
+/// time intensive.
+/// TODO: Codegen this as a kernel and run once at initialization
+static DenseElementsAttr foldFilterTransform(
+    ArrayRef<int64_t> shape, int64_t inputTileSize, int64_t kernelSize,
+    Type outputType, const float *G, bool isSplat, float splatValue,
+    DenseElementsAttr::iterator_range<APFloat> &input, Type elementType) {
+  const int &kh = shape[0];
+  const int &kw = shape[1];
+  const int &ic = shape[2];
+  const int &oc = shape[3];
+  const int64_t numElements = inputTileSize * inputTileSize * ic * oc;
+  SmallVector<APFloat> output(numElements, APFloat(0.0f));
+  for (int d0 = 0; d0 < inputTileSize; d0++) {
+    for (int d1 = 0; d1 < inputTileSize; d1++) {
+      for (int d2 = 0; d2 < ic; d2++) {
+        for (int d3 = 0; d3 < oc; d3++) {
+          APFloat accum(0.0f);
+          for (int d4 = 0; d4 < kernelSize; d4++) {
+            for (int d5 = 0; d5 < kernelSize; d5++) {
+              APFloat ival(splatValue);
+              if (!isSplat) {
+                ival = input[index(d4, d5, d2, d3, kh, kw, ic, oc)];
+              }
+              int idx0 = index(d0, d4, inputTileSize, kernelSize);
+              int idx1 = index(d1, d5, inputTileSize, kernelSize);
+              accum = accum + APFloat(G[idx0]) * ival * APFloat(G[idx1]);
+            }
+          }
+          int odx = index(d0, d1, d2, d3, inputTileSize, inputTileSize, ic, oc);
+          output[odx] = accum;
+        }
+      }
+    }
+  }
+  return DenseElementsAttr::get(outputType, output);
+}
+
+namespace {
+
+class FoldWinogradFilterTransform final
+    : public OpRewritePattern<linalg::Conv2DNhwcHwcfOp> {
+public:
+  using OpRewritePattern::OpRewritePattern;
+
+  LogicalResult matchAndRewrite(linalg::Conv2DNhwcHwcfOp convOp,
+                                PatternRewriter &rewriter) const override {
+    // Check that kernel size = 3x3
+    Value kernel = convOp.getInputs()[1];
+    auto kernelType = kernel.getType().cast<ShapedType>();
+    if (!kernelType)
+      return failure();
+    ArrayRef<int64_t> kernelShape = kernelType.getShape();
+    const int64_t kh = kernelShape[0];
+    const int64_t kw = kernelShape[1];
+    if ((kh != 3) || (kw != 3))
+      return failure();
+    const int64_t kernelSize = kh;
+    const int64_t inputTileSize = outputTileSize + kernelSize - 1;
+
+    DenseIntOrFPElementsAttr kernelAttr;
+    if (matchPattern(kernel, m_Constant(&kernelAttr))) {
```

Comment:
Nit: LLVM style is to have early exit. So 

```
if (!matchPattern(...)) {
  return failure();
}
.... // else part.
```
Removes one indentation.

Reply:
makes sense, will change